test/acceptance/audio-upload-composer-test.js

```diff
@@ -0,0 +1,174 @@
+import {
+  click,
+  find,
+  settled,
+  visit,
+  waitFor,
+  waitUntil,
+} from "@ember/test-helpers";
+import { test } from "qunit";
+import {
+  acceptance,
+  query,
+  visible,
+} from "discourse/tests/helpers/qunit-helpers";
+import I18n from "discourse-i18n";
+
+const uploadResponseData = {
+  id: 321,
+  url: "/uploads/default/original/1X/61fdf6fac415541560e2d86e495f94d4dd201a18.mp3",
+  original_filename: "recording.mp3",
+  filesize: 36864,
+  width: null,
+  height: null,
+  thumbnail_width: null,
+  thumbnail_height: null,
+  extension: "mp3",
+  short_url: "upload://dYSqLbGQHdjJFT40TIVt56uSIOs.mp3",
+  short_path: "/uploads/short-url/dYSqLbGQHdjJFT40TIVt56uSIOs.mp3",
+  retain_hours: null,
+  human_filesize: "36 KB",
+  dominant_color: null,
+  thumbnail: null,
+};
+
+acceptance("Audio Upload - Composer", function (needs) {
+  needs.user();
+  needs.settings({ authorized_extensions: "mp3" });
+  needs.pretender((server, helper) => {
+    server.post("/uploads.json", () => helper.response(uploadResponseData));
+  });
+
+  test("recording audio", async function (assert) {
+    await visit("/t/internationalization-localization/280");
+    await click("article#post_3 button.reply");
+
+    const buttonClass = ".d-editor-button-bar .composer_audio_upload";
+    assert.dom(buttonClass).exists("it adds a button to the composer toolbar");
+
+    await click(buttonClass);
+    assert.ok(visible(".d-modal"), "it pops up a modal");
+
+    // Default state
```

Comment:
@Arkshine Just for future reference, these states should be broken up into individual tests to make it easier to see points of failure. But not a big issue!

Reply:
Actually, I hesitated to break it into several parts because of how it works.
I agree with you. I hated having to comment to make it more readable, haha.